### PR TITLE
migrate azure_identity to `azure_core::error::Error`

### PR DESCRIPTION
**Pull request by ctaggart**
Changes:
- Removes `thiserror` errors & replaces with `azure_core::error::Error`
- Removes `identity::TokenCredential`
- `get_token` switched to `azure_core::error::Error`
- changes `Error::with_message` to accept a `FnOnce` like `ResultExt::with_context`.
- Adds `Error::message`, similar to `ResultExt::context`.
- refresh_token error handling now checks response status

### CODE REVIEW

**Comment by ctaggart** [comment]
I had to show myself what `format!` does under the covers.

``` rust
fn calc(s: impl Into<String>){
    s.into();
}

fn main() {
    calc(format!("Hello, world!"));
}
```

output from `cargo expand`:
``` rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
fn calc(s: impl Into<String>) {
    s.into();
}
fn main() {
    calc({
        let res = ::alloc::fmt::format(::core::fmt::Arguments::new_v1(&["Hello, world!"], &[]));
        res
    });
}
```

So yes, it is better to use a `|| format!`.

**Comment by johnbatty** [comment]
I'm happy with the current naming.  The `context`/`with_context` naming matches that of the corresponding `anyhow` functions, so this does have the benefit of being familiar to existing `anyhow` users.

File: sdk/identity/src/client_credentials_flow/mod.rs [hunk 1]
```diff
@@ -61,45 +62,26 @@ pub async fn perform(
...
+    })?;
 
     let response = client
         .post(url)
         .header("ContentType", "Application / WwwFormUrlEncoded")
         .body(encoded)
         .send()
         .await
-        .map_err(|e| ClientCredentialError::Request(Box::new(e)))?;
+        .map_kind(ErrorKind::Credential)?;
```
Comment (johnbatty) [inline sdk/identity/src/client_credentials_flow/mod.rs:75]: Perhaps this should be `ErrorKind::Io` if it fails to send?

File: sdk/identity/src/client_credentials_flow/mod.rs [hunk 2]
```diff
@@ -61,45 +62,26 @@ pub async fn perform(
...
 
-    if !response.status().is_success() {
-        return Err(ClientCredentialError::UnsuccessfulResponse(
-            response.status().as_u16(),
-            response.text().await.ok(),
-        ));
+    let rsp_status = response.status();
+    let rsp_body = response.bytes().await.map_kind(ErrorKind::Credential)?;
```
Comment (johnbatty) [inline sdk/identity/src/client_credentials_flow/mod.rs:78]: Again, perhaps this should be `ErrorKind::Io` if it fails to receive the body?

File: sdk/identity/src/device_code_flow/mod.rs [hunk 3]
```diff
@@ -39,44 +40,52 @@ where
...
-                client: Some(client),
-                tenant_id,
-                client_id: client_id.as_str().to_string(),
-            })
-        })
-        .map_err(|_| DeviceCodeError::InvalidResponseBody(s))?
+
+    let device_code_response = serde_json::from_slice::<DeviceCodePhaseOneResponse>(&rsp_body)
+    .with_context(
+        ErrorKind::Credential,
```
Comment (johnbatty) [inline sdk/identity/src/device_code_flow/mod.rs:72]: `ErrorKind::DataConversion`?

File: sdk/identity/src/device_code_flow/mod.rs [hunk 4]
```diff
@@ -39,44 +40,52 @@ where
...
         .await
-        .map_err(|e| DeviceCodeError::Request(Box::new(e)))?;
-
-    if !response.status().is_success() {
-        return Err(DeviceCodeError::UnsuccessfulResponse(
-            response.status().as_u16(),
-            response.text().await.ok(),
-        ));
+        .context(
+            ErrorKind::Credential,
```
Comment (johnbatty) [inline sdk/identity/src/device_code_flow/mod.rs:54]: `ErrorKind::Io`?

File: sdk/identity/src/device_code_flow/mod.rs [hunk 5]
```diff
@@ -39,44 +40,52 @@ where
...
+            "an error occurred when trying to make a request",
+        )?;
+
+    let rsp_status = response.status();
+    let rsp_body = response.bytes().await.context(
+        ErrorKind::Credential,
```
Comment (johnbatty) [inline sdk/identity/src/device_code_flow/mod.rs:60]: `ErrorKind::Io`?

**Review by johnbatty — COMMENTED** [review]
Mainly questions/clarifications about `ErrorKind` usage.

File: sdk/core/src/mock/mock_transaction.rs [hunk 6]
```diff
@@ -51,10 +51,10 @@ impl MockTransaction {
 
         if !path.exists() {
             if create_when_not_exist {
-                std::fs::create_dir_all(&path)
-                    .with_context(crate::error::ErrorKind::MockFramework, || {
-                        format!("cannot create transaction folder: {}", path.display())
-                    })?;
+                std::fs::create_dir_all(&path).context(
```
[the remark this answers is not in the document]
Comment (cataggar) [reply sdk/core/src/mock/mock_transaction.rs]: It is. I'll update.

File: sdk/identity/src/client_credentials_flow/mod.rs (hunk 1 above)
Comment (cataggar) [reply sdk/identity/src/client_credentials_flow/mod.rs:75]: Yeah, I wasn't sure if everything should wrap into `ErrorKind::Credential` or not in `azure_identity` crate. May be not.

File: sdk/identity/src/device_code_flow/mod.rs [hunk 7]
```diff
@@ -145,17 +151,18 @@ impl<'a> DeviceCodePhaseOneResponse<'a> {
                         .body(encoded)
                         .send()
                         .await
-                        .map_err(|e| DeviceCodeError::Request(Box::new(e)))
-                    {
+                        .context(
+                            ErrorKind::Credential,
```
Comment (johnbatty) [inline sdk/identity/src/device_code_flow/mod.rs:155]: `ErrorKind::Io`?

File: sdk/identity/src/device_code_flow/mod.rs [hunk 8]
```diff
@@ -145,17 +151,18 @@ impl<'a> DeviceCodePhaseOneResponse<'a> {
...
                         Err(error) => return Some((Err(error), NextState::Finish)),
                     };
 
-                    let result = match result
-                        .text()
-                        .await
-                        .map_err(|e| DeviceCodeError::Request(Box::new(e)))
-                    {
+                    let result = match result.text().await.context(
+                        ErrorKind::Credential,
```
Comment (johnbatty) [inline sdk/identity/src/device_code_flow/mod.rs:163]: `ErrorKind::Io`?

File: sdk/identity/src/refresh_token.rs [hunk 9]
```diff
@@ -44,37 +26,34 @@ pub async fn exchange(
...
         .post(url)
         .header("ContentType", "application/x-www-form-urlencoded")
         .body(encoded)
         .send()
         .await
-        .map_err(Error::Send)?
-        .text()
-        .await
-        .map_err(Error::Text)?;
+        .map_kind(ErrorKind::Credential)?;
```
Comment (johnbatty) [inline sdk/identity/src/refresh_token.rs:40]: `ErrorKind::Io`?

File: sdk/identity/src/refresh_token.rs [hunk 10]
```diff
@@ -44,37 +26,34 @@ pub async fn exchange(
...
 
-    debug!("refresh token response: {:?}", ret);
+    let rsp_status = rsp.status();
+    let rsp_body = rsp.bytes().await.map_kind(ErrorKind::Credential)?;
```
Comment (johnbatty) [inline sdk/identity/src/refresh_token.rs:43]: `ErrorKind::Io`?

File: sdk/identity/src/token_credentials/azure_cli_credentials.rs [hunk 11]
```diff
@@ -85,58 +66,51 @@ impl AzureCliCredential {
...
-                Err(AzureCliCredentialError::CommandFailed(output.to_string()))
+                Err(Error::with_message(ErrorKind::Credential, || {
+                    format!("az account get-access-token command failed: {output}")
+                }))
             }
             Err(e) => match e.kind() {
-                ErrorKind::NotFound => Err(AzureCliCredentialError::NotInstalled),
-                error_kind => Err(AzureCliCredentialError::UnknownError(error_kind)),
+                std::io::ErrorKind::NotFound => Err(Error::message(
+                    ErrorKind::Credential,
```
Comment (johnbatty) [inline sdk/identity/src/token_credentials/azure_cli_credentials.rs:83]: `ErrorKind::Other`?

File: sdk/identity/src/token_credentials/azure_cli_credentials.rs [hunk 12]
```diff
@@ -85,58 +66,51 @@ impl AzureCliCredential {
...
+                    "Azure CLI not installed",
+                )),
+                error_kind => Err(Error::with_message(ErrorKind::Credential, || {
```
Comment (johnbatty) [inline sdk/identity/src/token_credentials/azure_cli_credentials.rs:86]: `ErrorKind::Other`?

File: sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs [hunk 13]
```diff
@@ -108,7 +90,7 @@ impl TokenCredential for ImdsManagedIdentityCredential {
         }
 
         let msi_endpoint_url = Url::parse_with_params(&msi_endpoint, &query_items)
-            .map_err(ManagedIdentityCredentialError::MsiEndpointParseUrlError)?;
+            .context(ErrorKind::Credential, "error parsing url for MSI endpoint")?;
```
Comment (johnbatty) [inline sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs:93]: `ErrorKind::DataConversion`?

File: sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs [hunk 14]
```diff
@@ -121,16 +103,22 @@ impl TokenCredential for ImdsManagedIdentityCredential {
             .headers(headers)
             .send()
             .await
-            .map_err(ManagedIdentityCredentialError::SendError)?;
+            .map_kind(ErrorKind::Credential)?;
```
Comment (johnbatty) [inline sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs:106]: `ErrorKind::Io`?

File: sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs [hunk 15]
```diff
@@ -121,16 +103,22 @@ impl TokenCredential for ImdsManagedIdentityCredential {
...
 
         match response.status().as_u16() {
-            400 => Err(ManagedIdentityCredentialError::IdentityUnavailableError),
-            502 | 504 => Err(ManagedIdentityCredentialError::GatewayError),
+            400 => Err(Error::message(
+                ErrorKind::Credential,
```
Comment (johnbatty) [inline sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs:110]: `ErrorKind::HttpResponse`?

File: sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs [hunk 16]
```diff
@@ -121,16 +103,22 @@ impl TokenCredential for ImdsManagedIdentityCredential {
...
+                "the requested identity has not been assigned to this resource",
+            )),
+            502 | 504 => Err(Error::message(
+                ErrorKind::Credential,
```
Comment (johnbatty) [inline sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs:114]: `ErrorKind::HttpResponse`?

File: sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs [hunk 17]
```diff
@@ -121,16 +103,22 @@ impl TokenCredential for ImdsManagedIdentityCredential {
...
+                "the request failed due to a gateway error",
+            )),
             _ => {
                 let token_response = response
                     .json::<MsiTokenResponse>()
                     .await
-                    .map_err(ManagedIdentityCredentialError::DeserializeError)?;
+                    .map_kind(ErrorKind::Credential)?;
```
Comment (johnbatty) [inline sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs:121]: `ErrorKind::DataConversion`?

**Review by johnbatty — COMMENTED** [review]
More `ErrorKind` suggestions...  In my previous review I only gave a few as I wanted to know whether you agreed with the suggested changes.  These cover the whole PR.  Again, all are just suggestions so take or leave as you wish.

File: sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs (hunk 15 above)
Comment (cataggar) [reply sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs:110]: I'm going to leave this one. The error message is more specific.

**Review by johnbatty — APPROVED** [review]
Looks good - approved.

**Pull request merged**